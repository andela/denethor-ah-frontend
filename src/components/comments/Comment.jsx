import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import { TextAreaInput } from '../textarea';
import Button from '../Button';

export default class Comment extends Component {
  state = {
    comment: '',
    toastOptions: {
      hideProgressBar: true,
      closeButton: false,
      toastId: "commentToast",
      className: "toast-custom-style"
    }
  };

  handleOnChangeComment = ({ target: { value } }) => {
    this.setState({ comment: value });
  }

  handleOnClick = async (e) => {
    e.preventDefault();
    const { state: { comment, toastOptions }, props: { handleOnClick } } = this;

    if (!comment) return toast.error('Comment field cannot be empty', toastOptions);
    await handleOnClick(comment);
  }

  render() {
    const { 
      state: { comment }, props: { id }, handleOnChangeComment, handleOnClick 
    } = this;

    return (
      <div className='comment-create-field'>
        <TextAreaInput
          className={`article-comment-field ${id}`}
          placeHolder='Your message'
          value={comment}
          onChange={handleOnChangeComment}
        />
        <Button
          className='article-comment-field__button'
          type='submit'
          value='submit'
          onClick={handleOnClick}
        />
      </div>
    );
  }
}

Comment.propTypes = {
  handleOnClick: PropTypes.func,
  id: PropTypes.string
}
