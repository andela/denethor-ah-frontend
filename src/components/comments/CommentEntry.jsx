import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import EditComment from './editComment';
import EditCommentEntry from './editCommentEntry';
import { CommentImpressionButton } from "../impressionIcons";

export class CommentEntry extends Component {
  state = {
    mode: 'read'
  }

  toggleMode = (mode) => {
    this.setState({ mode })
  }

  render() {
    const {
      comment,
      comment: {
        commentBody,
        user: { firstname, lastname } = {},
        id, createdAt
      },
      allCommentsImpressions = [], handleOnClickLike, isLoggedIn
    } = this.props;
    const username = `${firstname} ${lastname}`
    const date = format(
      new Date(createdAt),
      'MMM D YYYY, hh:mm a'
    );

    const impression = allCommentsImpressions.length
      ? allCommentsImpressions.find(commentImpression => commentImpression.commentId === id)
      : {};
    const impressionCount = impression
    ? impression.count
    : 0;
  
    const { liked } = impression || {};

    const { mode } = this.state;

    return (
      <div className='comment-entry'>
        {
          mode === 'read'
            ?
            <div className='comment-text'>
              <p>{commentBody}</p>

                <div className='comment-text__info'>
              {firstname && <p className='comment-text__user-name'>{username}</p>}
                  <p className='comment-text__date'>{date}</p>
                </div>
            </div>
            :
            <EditCommentEntry  comment={comment} toggleMode={this.toggleMode} />
        }
        {firstname && (
          <div className='actions'>
            <div className='likes-count'>
              <CommentImpressionButton likeImpression={liked} onClick={() => console.log('a')} likeCount={impressionCount} />
            </div>
            <div className='comment__edit-comment__wrapper'>
              {isLoggedIn && <EditComment comment={comment} toggleMode={this.toggleMode} />}
            </div>
          </div>
        )}
      </div>
    );
  }
}

CommentEntry.propTypes = {
  comment: PropTypes.object,
  allCommentsImpressions: PropTypes.array,
  handleOnClickLike: PropTypes.func,
  likeImpression: PropTypes.bool,
  isLoggedIn: PropTypes.bool
};

const mapStateToProps = ({ auth: { isLoggedIn } }) => ({ isLoggedIn });

export default connect(mapStateToProps)(CommentEntry);
