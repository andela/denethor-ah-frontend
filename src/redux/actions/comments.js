import decodeJwt from 'jwt-decode';

import axios from '../../utils/axiosConfig';
import {
  ADD_COMMENT,
  REMOVE_COMMENT,
  ADD_COMMENT_FAILURE,
  GET_COMMENT_SUCCESS,
  COMMENT_IMPRESSION_SUCCESS,
  GET_COMMENT_LIKES_COUNT_SUCCESS
} from './types';

const BASE_URL = process.env.API_ROOT_URL;

export const addCommentRequest = () => ({
  type: ADD_COMMENT
});

export const getCommentSuccess = comments => ({
  type: GET_COMMENT_SUCCESS,
  comments
});

export const addCommentFailure = () => ({
  type: ADD_COMMENT_FAILURE
});

export const CommentImpressionSuccess = (commentImpression) => ({
  type: COMMENT_IMPRESSION_SUCCESS,
  commentImpression
})
export const getCommentslikesCountSuccess = (commentsLikes) => ({
  type: GET_COMMENT_LIKES_COUNT_SUCCESS,
  commentsLikes
})

export const getArticleComments = (articleId) => async (dispatch) => {
  const {
    data: {
      articleComments
    }
  } = await axios.get(`${BASE_URL}/articles/${articleId}/comments`);console.log(articleComments);
  dispatch(getCommentSuccess(articleComments));
};

export const addComment = ({
  articleId,
  commentBody
}) => async (dispatch) => {
  try {
    dispatch(addCommentRequest());
    await axios.post(`${BASE_URL}/articles/${articleId}/comments`, {
      commentBody
    });
    dispatch(getArticleComments(articleId));
  } catch (error) {
    dispatch(addCommentFailure())
    throw error;
  }
};

export const editComment = (
  articleId,
  commentId,
  commentBody
) => async (dispatch) => {
  try {
    dispatch(addCommentRequest());
    await axios.patch(`${BASE_URL}/articles/${articleId}/comments/${commentId}`, {
      commentBody
    });
    dispatch(getArticleComments(articleId));
  } catch (error) {
    dispatch(addCommentFailure())
    throw error;
  }
};

export const removeComment = id => (dispatch) => {
  dispatch({
    type: REMOVE_COMMENT,
    id
  });
};
export const getCommentImpression = async (commentId) => {
  const {
    data: {
      count,
      commentLikes
    }
  } = await axios.get(`${BASE_URL}/comments/${commentId}/likes`);
  const token = localStorage.getItem('token');
  const {
    id: userId
  } = decodeJwt(token);
  const liked = commentLikes.some(
    ({
      userId: likerId
    }) => likerId === userId);
  return {
    commentId,
    count,
    liked
  };
}
export const getAllCommentsImpression = (commentsIds) => async (dispatch) => {
  const commentsLikes = await Promise.all(
    commentsIds.map(async commentId => {
      const commentImpression = await getCommentImpression(commentId);
      return commentImpression;
    }),
  );
  dispatch(getCommentslikesCountSuccess(commentsLikes));
}

export const likeComment = (commentId) => async (dispatch) => {
  try {
    const {data: {status}} = await axios.post(`${BASE_URL}/comments/${commentId}/likes`);
    if (status === 'success') {
      dispatch(CommentImpressionSuccess(await getCommentImpression(commentId)));
    }
  } catch (error) {
    throw error;
  }
}
