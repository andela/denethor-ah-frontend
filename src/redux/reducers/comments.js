import {
  ADD_COMMENT,
  REMOVE_COMMENT,
  ADD_COMMENT_FAILURE,
  GET_COMMENT_SUCCESS,
  COMMENT_IMPRESSION_SUCCESS,
  GET_COMMENT_LIKES_COUNT_SUCCESS
} from '../actions/types';

export const commentReducerDefaultState = {
  comments: [],
  loading: false,
  commentsLikes: []
};

const commentReducer = (state = commentReducerDefaultState, action) => {
  let {
    commentsLikes
  } = state;
  const {
    commentImpression
  } = action;

  switch (action.type) {
    case ADD_COMMENT:
      return {
        ...state,
        loading: true
      }

    case GET_COMMENT_SUCCESS:
      return {
        ...state,
        comments: [...action.comments],
        loading: false
      };

    case ADD_COMMENT_FAILURE:
      return {
        ...state,
        loading: false
      }

    case REMOVE_COMMENT:
      return state.filter(({
        id
      }) => id !== action.id);

    case COMMENT_IMPRESSION_SUCCESS:
      {
        const filteredCommentsLikes = commentsLikes.filter(commentLike => 
          commentLike.commentId !== commentImpression.commentId
        );

        return {
          ...state,
          commentsLikes: [...filteredCommentsLikes, commentImpression]
        }
      }

    case GET_COMMENT_LIKES_COUNT_SUCCESS:
      return {
        ...state,
        commentsLikes: [...action.commentsLikes]
      }

    default:
      return state;
  }
};

export default commentReducer;