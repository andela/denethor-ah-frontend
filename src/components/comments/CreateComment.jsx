import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import { TextAreaInput } from '../textarea';
import Button from '../Button';
import { showLoginModal } from '../../redux/actions/auth';

export class Comment extends Component {
    constructor(props) {
        super(props);
        this.state = {
            commentBody: ''
        };
    }

    handleOnChangeComment = (event) => {
        this.setState({ commentBody: event.target.value });
    }

    handleOnClick = (e) => {
        e.preventDefault();
        if (!this.props.isLoggedIn) {
          this.props.dispatch(showLoginModal('Login'));
          return toast.error('You need to login to comment on this article');
        }
        const { articleId } = this.props;
        const { commentBody } = this.state;
        this.props.addComment({ articleId, commentBody })
        .then()
        .catch(error => {
        const { response } = error;
          if (response && response.status === 422){
            return toast.error(error.response.data.message.details[0].message);
          }
          if(response && response.status === 401){
            return toast.error('You need to login to comment on this article');
          }
          else{toast.error('You appear to be offline. Please check your Internet Connection');
        }
        });
        this.setState({
            commentBody: ''
          });
    }

    render() {
        const { commentBody }  = this.state;
        return (
            <div className='comment-create-field'>
            <TextAreaInput
                className='article-comment-field' 
                placeHolder='Your message'
                value={commentBody}
                onChange={this.handleOnChangeComment}
            />
            <Button
                className='article-comment-field__button' 
                type='submit' 
                value='submit'
                onClick={this.handleOnClick}
            />
            </div>
        );
    }
}

Comment.propTypes = {
    articleId: PropTypes.string.isRequired,
    addComment: PropTypes.func,
    dispatch: PropTypes.func,
    isLoggedIn: PropTypes.bool,
    value: PropTypes.string,
    showLoginModal: PropTypes.func
};

const mapStateToProps = (({ auth: { isLoggedIn } }) => ({ isLoggedIn }));

export default connect(mapStateToProps)(Comment);