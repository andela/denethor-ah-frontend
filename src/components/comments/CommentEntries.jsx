import React from "react";
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import './styles.scss';
import { LikeNumberGroup } from '../../components/likeNumberGroup';
import { HorizontalLine } from '../../components/horizontalLine';


const CommentEntries = ({ comments = [] }) => {
  const entries = comments.map((item, index) => {
    const { user: { firstname, lastname } } = item;
    const username = `${firstname} ${lastname}`
    const date = format(
      new Date(item.updatedAt),
      'MMM D YYYY, hh:mm a'
    );

    return (
      <div key={index} className='comment-entry'>
        <div className='comment-text'>
            <p>{item.commentBody}</p>
          <div className='comment-text__info'>
            <p className='comment-text__user-name'>{username}</p>
            <p className='comment-text__date'>{date}</p>
          </div>
        </div>
        <div className='actions'>
          <div className='likes-count'>
            {item.likesCount && <LikeNumberGroup likeCount={25} />}
          </div>
        </div>
      </div>
    );
  });

  const entriesExist = entries.length ? true : false;

  return (
    <div>
      {entriesExist && <div><HorizontalLine />Comments</div>}
      {entries}
    </div>
  );
}

CommentEntries.propTypes = {
  comments: PropTypes.array.isRequired
};

export default CommentEntries;
