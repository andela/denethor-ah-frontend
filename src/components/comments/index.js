import CommentEntries from './CommentEntries';
import CreateComment from './CreateComment';

export {CommentEntries, CreateComment};