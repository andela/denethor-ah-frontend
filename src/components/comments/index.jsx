import CommentEntries from './CommentEntries';

export { CommentEntries };