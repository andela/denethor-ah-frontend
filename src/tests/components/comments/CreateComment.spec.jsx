import React from 'react';
import { shallow } from 'enzyme';
import { toast } from 'react-toastify';
import { Comment } from '../../../components/comments/CreateComment';

jest.mock('../../../utils/socket.js');

jest.mock('react-toastify');
let wrapper, addComment;

it('Should render static component', () => {
  addComment = jest.fn();
  const props = { addComment, dispatch: () => {} }
  wrapper = shallow(<Comment {...props} />);
  expect(wrapper).toMatchSnapshot();
});

it('Should call handle functions onclick', async () => {
  addComment.mockResolvedValue('');
  wrapper.find('Button').simulate('click', { preventDefault: () => {} });
  expect(toast.error).toHaveBeenLastCalledWith('You need to login to comment on this article');

  wrapper.setProps({ isLoggedIn: true });
  wrapper.find('Button').simulate('click', { preventDefault: () => {} });
  expect(addComment).toHaveBeenCalled();
});
