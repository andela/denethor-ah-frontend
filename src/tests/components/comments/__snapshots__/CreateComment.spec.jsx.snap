// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render static component 1`] = `
<div
  className="comment-create-field"
>
  <TextAreaInput
    className="article-comment-field"
    onChange={[Function]}
    placeHolder="Your message"
    placeholder="Your message"
    value=""
  />
  <Button
    className="article-comment-field__button"
    isRequestSent={false}
    onClick={[Function]}
    type="submit"
    value="submit"
  />
</div>
`;
